lib/pop.md

```diff
@@ -0,0 +1,433 @@
+    program size and complexity to grow exponentially so as to distinguish
+    at every moment which entities in a deeply nested datastructure are
+    "only" prototypes, and which are already instances ready to be queried
+    but incapable of being composed anymore.
+
+  - Dynamic typing enables having all these fixed-points computations
+    where static typing would require dependent types or at least some
+    pretty fancy combination of subtyping and layered indexing.
+
+### POP versus other Nix extension mechanisms
+
+There are too many extension mechanisms in Nixpkgs already. Semantically,
+they are all isomorphic to each other and to Jsonnet's object system,
+with single-inheritance and `self: super:` extensions, yet
+they are all mutually incompatible, each with its own calling convention.
+They are insufficently documented, rarely consciously designed, often
+the reluctant product of necessity by programmers who are not well-versed
+in either or both of object systems and (pure lazy) functional programming
+(which is quite OK: not everyone has to be a programming language buff).
+
+The goal for POP would be to eventually replace all these systems with
+a single one, with a better more robust design, and hopefully feature parity
+and beyond (which it doesn't claim to have at the time).
+
+One notable feature that multiple inheritance might enable is that,
+if there are common aspects between packages for multiple languages,
+such as Haskell, OCaml, Python, Go, Gerbil Scheme, etc.,
+POP could modularly handle these aspects with shared prototypes,
+without each language reinventing their own hierarchy for whatever reason
+(e.g. absence of defaults, lozenge dependency conflicts, etc.).
+
+Also, a given package collection could independently mix and match
+alleles of multiple genes, such as the "stable" branch of all these but the
+"unstable" branch of all those, and the local checkout of those yet others.
+This could be done without having to maintain a lot of rigid yet fragile
+manual precedence lists for each combination of alleles used
+among an exponential number of possibilities.
+
+## Informal Types for POP
+
+Let's give informal types for POP, in a notional type system that has both
+subtyping guards and dependencies to some first- or second- class type indexes.
+These types are probably expressible in systems weaker than dependent types,
+but are not available in the typesystems of common programming languages,
+beyond the trivial unindexed monomorphic case.
+
+A `Proto A B` is a notional type for prototypes yielding an object of type `A`
+from a super object of type `B`, where `A` is a subtype of `B`. We write it thus:
+
+   type Proto = A: B: A B -> A | A <: B;
+
+An `Extension A B` is a notional type for prototype extensions yielding
+overrides to a super object of type `B` sufficient to turn it into
+an object of type `A`:
+
+   type Extension = A: B: Exists C: A B -> C | B // C <: A <: B;
```

Comment:
Perhaps explain a bit about the use of extensional types here or use `forall`

Reply:
Better?